### ISSUE TITLE
Animating Transitions

### ISSUE THREAD

**Issue by christoshrousis** [turbolinks/turbolinks#293]
This isn't an issue with Turbolinks perse, but I need some help with animating....

Let's say I write something like the following:

```javascript
$(document).on('turbolinks:click',function() {
  $('main').fadeOut('fast');
});
$(document).on('turbolinks:load',function() {
  $('main').fadeIn('fast');
});
```

Logically we're just trying to run an animation for fading out quickly, then fade in the content once it's loaded. Understandably, because the website is so fast with turbolinks, the page typically tends to flicker instead of running both animations to the full.

I'm assuming that my understanding of JS might be lacking here, but is there a way I can introduce an artificial wait time to turbo links fetch times to allow for the animation to run every time? Maybe I'm listening for the wrong events too I'm not 100% sure.... Any insight to this would be a great help. 

I tried googling some answers but they are pre-turbolinks 5 markup and they don't seem to translate well....

**Comment by murb** [comment]
It's a known issue. The problem is that Turbolinks replaces the DOM within the body relentlessly, and hence all animations attached to these elements.

There 's been [some experimentation with (Virtual)DOM diff'ing in this thread](https://github.com/turbolinks/turbolinks/issues/184), but none made it to become a real recommendation yet (although I [have used a follow-up solution I posted there several times](https://github.com/turbolinks/turbolinks/issues/184#issuecomment-265784720)).

Alternatively you could work around it by doing your own ajax call and push the state (use: `history.pushState({turbolinks: true, url: url}, '', url)`) and replace the main's content manually.

**Comment by kstratis** [comment]
Basically I've managed to get page transitions working like this:

1. Download [animate.css](https://raw.github.com/daneden/animate.css/master/animate.css) and put it in your head tag.
2. Make a new file `javascripts/turbolinks_transitions.js` with the contents below:

```javascript
$(document).on('turbolinks:load', function(){
  $('.div-to-animate').addClass('animated fadeIn');
});
```
and make sure it is loaded in `application.js` like this:
    
    ...
    //= require turbolinks
    //= require turbolinks_transitions
    ... 

On each visit turbolinks swaps the document body so 'animated' and 'fadeIn'
are automatically removed and don't add up. Thus you don't have to worry about removing classes. You just need to listen to this one event.

**Comment by neohunter** [comment]
I'd this problem today, the loading appears instantly because turbolinks cache, it renders the page before the ajax call is made, breaking the animation.

In order to get a loading effect working we must disable it:

```
<head>
....
  <meta name="turbolinks-cache-control" content="no-cache">
```

Then we can do something like: (using animated.css)

```
$(document).on('turbolinks:click', function(){
  $('.page-content')
    .addClass('animated fadeOut')
    .off('webkitAnimationEnd oanimationend msAnimationEnd animationend')
});


$(document).on('turbolinks:load', function(event){
    // if call was fired by turbolinks
    if (event.originalEvent.data.timing.visitStart) { 
      $('.page-content')
        .addClass('animated fadeIn')
        .one('webkitAnimationEnd oanimationend msAnimationEnd animationend', function(){
          $('.page-content').removeClass('animated');
        });
    }else{
      $('.page-content').removeClass('hide')
    }  
...
```

With that, the transition is nice and smooth.

**Comment by david-treblig** [comment]
I would argue that page transitions today really help communicating navigation and that there should be a clearly defined callback / event to use. Something that uses promises or something like that would be really nice so that we can delay the cycle as we want.

**Comment by jondavidchristopher** [comment]
@david-treblig totally agree. It can also make a 200ms page load feel instant if you have a 300ms animation.

**Comment by luislezcair** [comment]
I agree. I've been trying to add fade in animations with cached pages but there are still cases where I cannot control (like visiting a page for the first time).

**Comment by lasoiveille** [comment]
Not necessarily a "fix," but there is a project that's eased these headaches for me when doing transitions with turbolinks: [https://github.com/jonhue/turbolinks-animate](https://github.com/jonhue/turbolinks-animate)

**Comment by david-treblig** [comment]
@lasoiveillw Turbolinks animate : It is bundled like a gem for rails, but my guess is picking only the npm package would work for any web project outside of the framework? Or is there any obscure modification allowed only by the fact of bein a gem?

**Comment by david-treblig** [comment]
Forget my comment, I read too fast. The link you provided is only a gem wrapper for this one : https://github.com/jonhue/turbolinks-animate.js 👍👍

**Comment by lasoiveille** [comment]
Ah, I should have checked which link I gave lol I thought it was the js one. But now both links are here for the future 👍

**Comment by Geestig** [comment]
+1 would love a way to delay the actual 'visit' in order to have enough time to make everything fade out before starting the new page from being faded in.

**Comment by tedw** [comment]
Just came across this fwiw https://github.com/gmrchk/swup

via https://css-tricks.com/page-transitions-for-everyone/

**Comment by domchristie** [comment]
I've been experimenting with this and have made a start on a Turbolinks version which allows rendering to be deferred—useful for waiting for exit animations to end.

To use, add a `turbolinks:before-render` event listener, prevent the default, then call `event.data.render()` when ready:

```js
addEventListener('turbolinks:before-render', function (event) {
  event.preventDefault()
  // …
  // Wait for animations to end
  // …
  event.data.render()
})
```

I've not tested it out much, and I'm thinking it might require a bit of work to cover all visit cases (like restorations), but I thought I'd put it out there. The branch is here: https://github.com/domchristie/turbolinks/tree/defer_render, with the [corresponding diff](https://github.com/turbolinks/turbolinks/compare/master...domchristie:defer_render).

**Comment by domchristie** [comment]
… and if anyone's interested, a lightweight proof of concept: https://scarce-incredible-dormouse.glitch.me/
([Source](https://glitch.com/edit/#!/scarce-incredible-dormouse))

**Comment by christoshrousis** [comment]
A sensible solution in my eyes, have you sent a PR? What's outstanding to get it ready for PR?

(Can't believe I posted this in 2017, I know so much more JS now 😂)

**Comment by stuartcusackie** [comment]
Just wanted to add that CSS transitions are problematic in general with Turbolinks. I can't get [AOS](https://michalsnik.github.io/aos/) to play well with Turbolinks.

Swup on the otherhand works fine with AOS, but I would prefer to stick with Turbolinks.

**Comment by stuartcusackie** [comment]
I wonder if this will work with CSS transitions in general? A lot of my CSS transitions don't fire because the page swaps out too quickly, for example, my dropdowns usually fade out before a redirect. I've had to disable TL caching to keep this behaviour. 

In short, it would be good it Turbolinks would wait for all CSS transitions to finish, or if we had an option to delay things. Swup does a good job of this but I would like to stick with Turbolinks.

**Comment by domchristie** [comment]
> A sensible solution in my eyes, have you sent a PR? What's outstanding to get it ready for PR?

Not yet. I've still got to write some tests. There are some upcoming changes to Turbolinks, which might cause some merge conflicts, but I suppose the implementation might be worth sharing anyway.

> I wonder if this will work with CSS transitions in general?

Yep! It's might be too much to expect Turbolinks to wait for any transitions to end. (Potentially they could never end!) But for a limited set of elements this should work well. In the demo I posted, I create an array of promises that resolve when their animations end (see below). This could be adapted for transitions too.

```js
function animationsEnd (selector) {
  const elements = document.querySelectorAll(selector)
  const promises = []

  elements.forEach((element) => {
    const promise = new Promise((resolve) => {
      function listener () {
        element.removeEventListener('animationend', listener)
        resolve()
      }
      element.addEventListener('animationend', listener)
    })
    promises.push(promise)
  })

  return Promise.all(promises)
}
```

**Comment by osbre** [comment]
Updated version using [animate.css](https://github.com/daneden/animate.css/)

```js
document.addEventListener('turbolinks:load', (event) => {
    document.body.classList.add('animate__animated');
    document.body.classList.add('animate__fadeIn');
})

```
